packages/rest/src/router/regexp-router.ts

```diff
@@ -0,0 +1,94 @@
+// Copyright IBM Corp. 2017, 2018. All Rights Reserved.
+// Node module: @loopback/rest
+// This file is licensed under the MIT License.
+// License text available at https://opensource.org/licenses/MIT
+
+import {RouteEntry, createResolvedRoute, ResolvedRoute} from './routing-table';
+import {Request, PathParameterValues} from '../types';
+import {inspect} from 'util';
+import {compareRoute} from './route-sort';
+import pathToRegExp = require('path-to-regexp');
+import {BaseRouter} from './router-base';
+import {toExpressPath} from './openapi-path';
+
+const debug = require('debug')('loopback:rest:router:regexp');
+
+/**
+ * Route entry with path-to-regexp
+ */
+interface RegExpRouteEntry extends RouteEntry {
+  regexp?: RegExp;
+  keys?: pathToRegExp.Key[];
+}
+
+/**
+ * Router implementation based on regexp matching
+ */
+export class RegExpRouter extends BaseRouter {
+  private routes: RegExpRouteEntry[] = [];
+
+  // Sort the routes based on their paths and variables
+  private _sorted: boolean;
+  private _sort() {
+    if (!this._sorted) {
+      this.routes.sort(compareRoute);
+      this._sorted = true;
+    }
+  }
+
+  add(route: RouteEntry) {
+    if (super.add(route)) return true;
```

Comment:
I find this design brittle, as it's relying on developers implementing custom Router classes to not forget to handle this special case.

Have you considered using Template Design Pattern?

```ts
export class BaseRouter {
  add(route: RouteEntry) {
     if (getPathVariables(route.path)) {
       this.addTemplatedRoute(route);
     } else {
       const key = this.getKeyForRoute(route);
       this.simpleRoutes[key] = route;
     }
  }

  protected addTemplatedRoute(route: RouteEntry) {
    throw new Error('This method is abstract');
  }

  // ...
}

export class RegExpRouter extends BaseRouter {
   private routes: RegExpRouteEntry[] = [];

  addTemplatedRoute(route: RouteEntry) {
     const entry = route as RegExpRouteEntry;
     // ...
  }

  // ...
}
```

Similarly for find/list.